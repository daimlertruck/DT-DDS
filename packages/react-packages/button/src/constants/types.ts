import { ComponentSize } from '@dt-ui/react-core';

export type ButtonVariant = 'solid' | 'outlined' | 'text';

export type ButtonColor = 'primary' | 'error' | 'secondary';

export type ButtonAppearance = ButtonColor | 'contrast';

export type ButtonSize = Extract<ComponentSize, 'medium' | 'large' | 'small'>;
