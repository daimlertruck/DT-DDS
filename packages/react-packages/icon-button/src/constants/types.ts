export type IconButtonVariant = 'contrast' | 'default';
