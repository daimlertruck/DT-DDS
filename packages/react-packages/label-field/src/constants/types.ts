export type FieldScale = 'compact' | 'standard';
