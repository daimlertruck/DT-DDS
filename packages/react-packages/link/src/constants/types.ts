export type LinkColor = 'accent' | 'secondary';
export type LinkSize = 'small' | 'medium' | 'large';
export type LinkVariant = 'standalone' | 'inline';
