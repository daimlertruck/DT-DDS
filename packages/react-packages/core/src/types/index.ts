export * from './baseProps';
export * from './componentSize';
export * from './fontFace';
export * from './orientation';
export * from './theme';
