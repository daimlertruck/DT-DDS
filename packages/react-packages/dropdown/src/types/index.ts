export type DropdownOptionValue = {
  text?: string;
  value: string;
};
