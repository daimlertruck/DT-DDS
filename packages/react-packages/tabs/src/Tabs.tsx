import { Box } from '@dt-ui/react-box';
import { BaseProps, useDebounceResize } from '@dt-ui/react-core';
import { Icon } from '@dt-ui/react-icon';
import {
  Children,
  cloneElement,
  ReactElement,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';

import { TabItem, TabItemProps } from './components';
import { TabsStyled } from './Tabs.styled';
import { Variant } from './types';

const SCROLL_MOVEMENT = 120;
const SCROLL_OFFSET = 0.5;

export interface TabsProps extends BaseProps {
  activeTab: string;
  variant?: Variant;
}

export const Tabs = ({
  children,
  dataTestId,
  style,
  activeTab,
  variant = 'boxed',
}: TabsProps) => {
  const ref = useRef<HTMLDivElement>(null);
  const [isLeftSideOverflowing, setIsLeftSideOverflowing] = useState(false);
  const [isRightSideOverflowing, setIsRightSideOverflowing] = useState(false);

  const clonedChildren = useMemo(
    () =>
      Children.map(children as ReactElement<TabItemProps>, (child) => {
        return (
          child &&
          cloneElement(child, {
            ...child.props,
            activeTab,
            variant,
          })
        );
      }),
    [children, activeTab, variant]
  );

  const handleScroll = (scrollOffset: number) => {
    if (!ref || !ref.current) return;

    const left = ref.current.scrollLeft + scrollOffset;
    ref.current.scroll({
      left,
      behavior: 'smooth',
    });

    calculateIfIsOverflowing(left);
  };

  const calculateIfIsOverflowing = (currentScrollLeft?: number) => {
    if (!ref || !ref.current) return;

    const scrollLeft = currentScrollLeft ?? ref.current.scrollLeft;

    setIsLeftSideOverflowing(scrollLeft > 0);
    setIsRightSideOverflowing(
      ref.current.clientWidth + scrollLeft + SCROLL_OFFSET <
        ref.current.scrollWidth
    );
  };

  useEffect(() => calculateIfIsOverflowing(), []);

  useDebounceResize(() => calculateIfIsOverflowing());

  return (
    <Box style={{ flexDirection: 'row' }}>
      <Icon
        code='keyboard_arrow_left'
        dataTestId='left-arrow'
        onClick={() => handleScroll(-SCROLL_MOVEMENT)}
        style={{
          visibility: isLeftSideOverflowing ? 'visible' : 'hidden',
        }}
      />
      <TabsStyled
        data-testid={dataTestId ? dataTestId : 'tabs'}
        ref={ref}
        role='tablist'
        style={style}
      >
        {clonedChildren}
      </TabsStyled>
      <Icon
        code='keyboard_arrow_right'
        dataTestId='right-arrow'
        onClick={() => handleScroll(SCROLL_MOVEMENT)}
        style={{
          visibility: isRightSideOverflowing ? 'visible' : 'hidden',
        }}
      />
    </Box>
  );
};

Tabs.Item = TabItem;
