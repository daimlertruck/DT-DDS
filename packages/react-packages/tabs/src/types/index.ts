export type Variant = 'boxed' | 'book';
