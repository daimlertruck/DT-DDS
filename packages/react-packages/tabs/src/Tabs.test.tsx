import { withProviders } from '@dt-dds/react-core';
import { Icon } from '@dt-dds/react-icon';
import { fireEvent, render, screen } from '@testing-library/react';
import { Fragment } from 'react';

import { Tabs } from './Tabs';
import { Variant } from './types';

const handleChangeMock = jest.fn();

describe('<Tabs /> component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it.each`
    variant
    ${'default'}
    ${'contained'}
  `(
    'should match the snapshot by displaying $variant variant',
    ({ variant }) => {
      const { container } = renderComponent(handleChangeMock, variant);

      expect(container).toMatchSnapshot();
    }
  );

  it('should display side icons when scrollable', () => {
    expect(screen.queryByTestId('left-arrow')).toBeNull();
    expect(screen.queryByTestId('right-arrow')).toBeNull();

    Object.defineProperty(HTMLElement.prototype, 'scrollLeft', {
      writable: true,
      value: 10,
    });

    Object.defineProperty(HTMLElement.prototype, 'clientWidth', {
      writable: true,
      value: 10,
    });

    Object.defineProperty(HTMLElement.prototype, 'scrollWidth', {
      writable: true,
      value: 100,
    });

    renderComponent(handleChangeMock);

    expect(screen.queryByTestId('left-arrow')).toBeVisible();
    expect(screen.queryByTestId('right-arrow')).toBeVisible();
  });

  it.each`
    scroll
    ${'left'}
    ${'right'}
  `('should handle scroll after $scroll side icon click', ({ scroll }) => {
    const scrollMock = jest.fn();

    Object.defineProperty(HTMLElement.prototype, 'scroll', {
      writable: true,
      value: scrollMock,
    });

    renderComponent(handleChangeMock);

    const scrollButton = screen.getByTestId(`${scroll}-arrow`);

    fireEvent.click(scrollButton);

    expect(scrollMock).toHaveBeenCalled();
  });

  it('should trigger handle change after item click', () => {
    const { getByTestId } = renderComponent(handleChangeMock);

    fireEvent.click(getByTestId('tab-item-1'));

    expect(handleChangeMock).toHaveBeenCalledWith(1);
  });
});

const renderComponent = (
  handleChange: () => void,
  variant: Variant = 'default'
) => {
  const ProvidedFragment = withProviders(Fragment);

  return render(
    <ProvidedFragment>
      <Tabs activeTab={0} handleChange={handleChange} variant={variant}>
        <Tabs.Item label='Tab 1' />
        <Tabs.Item label='Tab 2' />
        <Tabs.Item icon={<Icon code='menu' />} label='Tab 3' />
        <Tabs.Item isDisabled label='Tab 4' />
      </Tabs>
    </ProvidedFragment>
  );
};
