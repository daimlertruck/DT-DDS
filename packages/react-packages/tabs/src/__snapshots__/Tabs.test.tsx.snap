// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> component should match the snapshot by displaying contained variant 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-2 {
  font-family: DTUI-icons-outlined;
  font-weight: normal;
  font-style: normal;
  font-size: 24px;
  display: inline-block;
  line-height: 1;
  text-transform: none;
  letter-spacing: normal;
  word-wrap: normal;
  white-space: nowrap;
  font-variation-settings: 'FILL' 0;
  color: #454545;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -moz-osx-font-smoothing: grayscale;
  font-feature-settings: 'liga';
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  overflow: hidden;
  width: 100%;
}

.emotion-6 {
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 700;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
  background-color: #FFFFFF;
  color: #00677F;
  padding: 12px 32px;
  cursor: default;
  white-space: nowrap;
  gap: 8px;
}

.emotion-8 {
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
  background-color: transparent;
  color: #00677F;
  padding: 12px 32px;
  cursor: pointer;
  white-space: nowrap;
  gap: 8px;
}

.emotion-8:hover {
  background-color: #CCE3EB;
}

.emotion-12 {
  font-family: DTUI-icons-outlined;
  font-weight: normal;
  font-style: normal;
  font-size: 24px;
  display: inline-block;
  line-height: 1;
  text-transform: none;
  letter-spacing: normal;
  word-wrap: normal;
  white-space: nowrap;
  font-variation-settings: 'FILL' 0;
  color: #454545;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -moz-osx-font-smoothing: grayscale;
  font-feature-settings: 'liga';
}

.emotion-14 {
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
  background-color: transparent;
  color: #A3A3A3;
  padding: 12px 32px;
  cursor: not-allowed;
  white-space: nowrap;
  gap: 8px;
}

<div>
  <div
    class="emotion-0 emotion-1"
    style="flex-direction: row;"
  >
    <i
      class="emotion-2 emotion-3"
      color="#454545"
      data-testid="left-arrow"
      style="display: none;"
    >
      keyboard_arrow_left
    </i>
    <div
      class="emotion-4 emotion-5"
      data-testid="tabs"
      role="tablist"
    >
      <button
        class="emotion-6 emotion-7"
        data-testid="tab-item-0"
        role="tab"
        tabindex="0"
      >
        Tab 1
      </button>
      <button
        class="emotion-8 emotion-7"
        data-testid="tab-item-1"
        role="tab"
        tabindex="-1"
      >
        Tab 2
      </button>
      <button
        class="emotion-8 emotion-7"
        data-testid="tab-item-2"
        role="tab"
        tabindex="-1"
      >
        <i
          class="emotion-12 emotion-3"
          color="#454545"
          data-testid="icon"
        >
          menu
        </i>
        Tab 3
      </button>
      <button
        class="emotion-14 emotion-7"
        data-testid="tab-item-3"
        disabled=""
        role="tab"
        tabindex="-1"
      >
        Tab 4
      </button>
    </div>
    <i
      class="emotion-2 emotion-3"
      color="#454545"
      data-testid="right-arrow"
      style="display: none;"
    >
      keyboard_arrow_right
    </i>
  </div>
</div>
`;

exports[`<Tabs /> component should match the snapshot by displaying default variant 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-2 {
  font-family: DTUI-icons-outlined;
  font-weight: normal;
  font-style: normal;
  font-size: 24px;
  display: inline-block;
  line-height: 1;
  text-transform: none;
  letter-spacing: normal;
  word-wrap: normal;
  white-space: nowrap;
  font-variation-settings: 'FILL' 0;
  color: #454545;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -moz-osx-font-smoothing: grayscale;
  font-feature-settings: 'liga';
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  overflow: hidden;
  width: 100%;
  margin-bottom: -1px;
}

.emotion-6 {
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 700;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
  background-color: transparent;
  border-bottom: 2px solid #00677F;
  color: #00677F;
  padding: 12px 32px;
  cursor: default;
  white-space: nowrap;
  gap: 8px;
}

.emotion-8 {
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
  background-color: transparent;
  false color: #00677F;
  padding: 12px 32px;
  cursor: pointer;
  white-space: nowrap;
  gap: 8px;
}

.emotion-8:hover {
  background-color: #CCE3EB;
}

.emotion-12 {
  font-family: DTUI-icons-outlined;
  font-weight: normal;
  font-style: normal;
  font-size: 24px;
  display: inline-block;
  line-height: 1;
  text-transform: none;
  letter-spacing: normal;
  word-wrap: normal;
  white-space: nowrap;
  font-variation-settings: 'FILL' 0;
  color: #454545;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -moz-osx-font-smoothing: grayscale;
  font-feature-settings: 'liga';
}

.emotion-14 {
  border: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
  background-color: transparent;
  false color: #A3A3A3;
  padding: 12px 32px;
  cursor: not-allowed;
  white-space: nowrap;
  gap: 8px;
}

<div>
  <div
    class="emotion-0 emotion-1"
    style="flex-direction: row; border-bottom: 1px solid #DEDEDE;"
  >
    <i
      class="emotion-2 emotion-3"
      color="#454545"
      data-testid="left-arrow"
      style="display: none;"
    >
      keyboard_arrow_left
    </i>
    <div
      class="emotion-4 emotion-5"
      data-testid="tabs"
      role="tablist"
    >
      <button
        class="emotion-6 emotion-7"
        data-testid="tab-item-0"
        role="tab"
        tabindex="0"
      >
        Tab 1
      </button>
      <button
        class="emotion-8 emotion-7"
        data-testid="tab-item-1"
        role="tab"
        tabindex="-1"
      >
        Tab 2
      </button>
      <button
        class="emotion-8 emotion-7"
        data-testid="tab-item-2"
        role="tab"
        tabindex="-1"
      >
        <i
          class="emotion-12 emotion-3"
          color="#454545"
          data-testid="icon"
        >
          menu
        </i>
        Tab 3
      </button>
      <button
        class="emotion-14 emotion-7"
        data-testid="tab-item-3"
        disabled=""
        role="tab"
        tabindex="-1"
      >
        Tab 4
      </button>
    </div>
    <i
      class="emotion-2 emotion-3"
      color="#454545"
      data-testid="right-arrow"
      style="display: none;"
    >
      keyboard_arrow_right
    </i>
  </div>
</div>
`;
